src/main/java/com/endava/internship/collections/StudentList.java

```diff
@@ -1,145 +1,332 @@
+        }
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public Student set(int i, Student student) {
-        //TODO
-        return null;
+    public T set(int i, T t) {
+        if (i >= 0 && i < size) {
+            return (T) (elements[i] = t);
+        } else {
+            throw new IndexOutOfBoundsException("Insertion index was out of range. Must be non-negative and less than or equal to size");
+        }
     }
 
     @Override
-    public void add(int i, Student student) {
-        //TODO
+    public void add(int i, T t) {
+        if (i < 0 || i > size) {
+            throw new IndexOutOfBoundsException("Insertion index was out of range. Must be non-negative and less than or equal to size");
+        }
+        if (size == elements.length) {
+            resizeArray();
+        }
+        if (size - i >= 0) {
+            System.arraycopy(elements, i, elements, i + 1, size - i);
+        }
+        elements[i] = t;
+        size++;
     }
 
     @Override
-    public Student remove(int i) {
-        //TODO
-        return null;
+    public T remove(int i) {
+        if (i < 0 || i >= size) {
+            throw new IndexOutOfBoundsException("Insertion index was out of range. Must be non-negative and less than or equal to size");
+        }
+        T elementToRemove = this.get(i);
+        System.arraycopy(elements, i + 1, elements, i, size() - i);
+        elements[size] = null;
+        size--;
+        return elementToRemove;
     }
 
     @Override
-    public int indexOf(Object o) {
-        //TODO
-        return 0;
+    public int indexOf(Object t) {
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(t, elements[i])) {
+                return i;
+            }
+        }
+        return -1;
     }
 
     @Override
-    public int lastIndexOf(Object o) {
-        //TODO
-        return 0;
+    public int lastIndexOf(Object t) {
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(t, elements[i])) {
+                return i;
+            }
+        }
+        return -1;
     }
 
     @Override
-    public ListIterator<Student> listIterator() {
-        //TODO
-        return null;
+    public ListIterator<T> listIterator() {
+        return new ListIter(0);
     }
 
     @Override
-    public ListIterator<Student> listIterator(int i) {
-        //TODO
-        return null;
+    public ListIterator<T> listIterator(int i) {
+        return new ListIter(i);
+    }
+
+    private class ListIter extends Iter implements ListIterator<T> {
+        ListIter(int index) {
+            super();
+            currentElement = index;
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            return currentElement > 0;
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public T previous() {
+            --currentElement;
```

Comment:
shouldn't  you first check if it hasPrevious ?

Reply:
No it will be checked in the "while"

Reply:
Checked if hasPrevious. Done